File: 24.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

        
class Solution:
    def swapNode(self, node, prevNode):
        if node.next != None:
            nodeA = node
            prevNode.next = node.next
            nodeA.next = node.next.next
            prevNode.next.next = nodeA
            if nodeA.next != None and nodeA.next.next != None:
                self.swapNode(node.next, node)

    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None:
            return None

        if head.next == None:
            return head

        formerHead = head
        head = head.next
        formerHead.next = head.next
        head.next = formerHead

        if formerHead.next != None and formerHead.next.next != None:
            self.swapNode(formerHead.next, formerHead)

        return head

        


solution = Solution()
solution.swapPairs(ListNode(1, ListNode(2, ListNode(3, ListNode(4)))))

File: 24.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function swapPairs(head: ListNode | null): ListNode | null {
  if (!head || !head.next) return head;

  let dummyNode: ListNode = new ListNode(0, head);

  let nodeBeforePair = dummyNode;
  let nodeAfterPair: ListNode | null = null;

  let current = dummyNode.next;

  while (current && current.next) {
    nodeAfterPair = current.next.next;
    let nodeA = current;
    let nodeB = current.next;

    nodeBeforePair.next = nodeB;
    nodeB.next = nodeA;
    nodeA.next = nodeAfterPair;

    current = nodeAfterPair;
    nodeBeforePair = nodeA;
  }

  return dummyNode.next;
}
